### PR TITLE
nomad exec: Allow customizing environment variables

### PR DESCRIPTION
**Pull request by notnoop**
When execing into a task, it's useful to be able to set some special
environment variables, namely `TERM` and language environment variable.

In this PR, we expose `env` parameter that takes an array of `KEY=VALUE`
values that gets thread all the way through to executors exec handling.

The main purpose is to allow configuring `TERM` for purposes of web ui,
as the appropriate value would depend on the terminal library and not
necessarily user friendly.

This PR does set `TERM` in the CLI case to user.  The aim is to ease interactive development with text-based UIs (e.g. top/htop, tmux, vim, etc).  Other language and profile environment variables (e.g. `LANG`, `LC_...`) might be useful, but the target may not always support languages set by user and the default languages in target usually are sensible for that environment.

CLI
----------

This PR doesn't expose this configuration to CLI due to some
complexities:

* It's not as valuable to set environment variables there
(i.e. users can export their environment variables before running
commands)
* The `-e` flag is already used for escape character, and I'm unsure if
how to best proceed; I can see the appeal of repurposing `-e` to
env-variables as it's the most common thing
* We use the standard library `flags` package that doesn't support
string array flag to allow for multiple environment variables nicely.

As such, I've punted on supporting the CLI for another time as the need
arises, but special cased `TERM` as it's useful on its own now.

External drivers and compatibility
----------

This PR makes some changes to the exec endpoint in drivers.  These
changes are binary backward compatible: existing drivers continue to run
against latest nomad binary as the grpc handler ignores the new `env`
field.  The drivers continue to compile as normal against the version
they vendored or depended on.

Though, when plugin authors upgrade the dependency they will need to
update code to either handle the `env` field or explicitly ignore it.

### CODE REVIEW

File: api/allocations.go [hunk 1]
```diff
@@ -84,7 +84,7 @@ func (a *Allocations) Info(allocID string, q *QueryOptions) (*Allocation, *Query
 //
 // The call blocks until command terminates (or an error occurs), and returns the exit code.
 func (a *Allocations) Exec(ctx context.Context,
-	alloc *Allocation, task string, tty bool, command []string,
+	alloc *Allocation, task string, tty bool, command, envs []string,
```
Comment (cjonesy) [inline api/allocations.go]:
```suggestion
	alloc *Allocation, task string, tty bool, command []string, envs []string,
```

File: command/alloc_exec.go [hunk 2]
```diff
@@ -230,7 +234,7 @@ func (l *AllocExecCommand) Run(args []string) int {
 
 // execImpl invokes the Alloc Exec api call, it also prepares and restores terminal states as necessary.
 func (l *AllocExecCommand) execImpl(client *api.Client, alloc *api.Allocation, task string, tty bool,
-	command []string, escapeChar string, stdin io.Reader, stdout, stderr io.WriteCloser) (int, error) {
+	command, envs []string, escapeChar string, stdin io.Reader, stdout, stderr io.WriteCloser) (int, error) {
```
Comment (cjonesy) [inline command/alloc_exec.go]:
```suggestion
	command []string, envs []string, escapeChar string, stdin io.Reader, stdout, stderr io.WriteCloser) (int, error) {
```

File: api/allocations.go (hunk 1 above)
Comment (schmichael) [reply api/allocations.go:87]: Go allows skipping redundant type declaration when multiple func parameters have the same type, so the code is ok as-is.

File: command/alloc_exec.go (hunk 2 above)
Comment (schmichael) [reply command/alloc_exec.go:237]: Go allows skipping redundant type declaration when multiple func parameters have the same type, so the code is ok as-is.

File: client/structs/structs.go [hunk 3]
```diff
@@ -175,6 +175,9 @@ type AllocExecRequest struct {
 	// Cmd is the command to be executed
 	Cmd []string
 
+	// Envs is the environment variables to be set
+	Envs []string
```
Comment (schmichael) [inline client/structs/structs.go:179]: Sorry to be really pedantic, but this kind of field is called `Env`, `Envvars`, or `Environ` elsewhere. "Enviornment" being singluar, but "Environment variables" being plural.

**Comment by notnoop** [comment]
This doesn't seem like a high priority item and we don't seem to require it.  Considering it's a backward incompatible change, I'll close it and consider resurrecting it in 0.12 development cycle if we want it.

**Pull request closed without merging**